src/FsAutoComplete/CodeFixes/AdjustConstant.fs

```diff
@@ -0,0 +1,1670 @@
+      $"%A{constant.Sign} %A{intPart}.%A{decPart}e%A{expPart}%A{suffix} (%s{format}) (%A{full}, %A{constant.Value})"
+
+    mkFix doc data [||]
+
+  /// Separates digit groups with `_`.
+  let separateDigitGroupsFix doc (lineStr: String) (constant: FloatConstant) =
+    let text = constant.Range.SpanIn(lineStr)
+
+    if text.Contains '_' then
+      []
+    else
+      let edits =
+        [| if constant.IntRange.Length > 3 then
+             let range = constant.IntRange.ToRangeInside constant.Range
+             let n = range.SpanIn(lineStr).ToString()
+
+             { Range = range
+               NewText = DigitGroup.addSeparator n 3 DigitGroup.RightToLeft }
+           if constant.DecimalRange.Length > 3 then
+             let range = constant.DecimalRange.ToRangeInside constant.Range
+             let n = range.SpanIn(lineStr).ToString()
+
+             { Range = range
+               NewText = DigitGroup.addSeparator n 3 DigitGroup.LeftToRight }
+           if constant.ExponentRange.Length > 3 then
+             let range = constant.ExponentRange.ToRangeInside constant.Range
+             let n = range.SpanIn(lineStr).ToString()
+
+             { Range = range
+               NewText = DigitGroup.addSeparator n 3 DigitGroup.RightToLeft } |]
+
+      match edits with
+      | [||] -> []
+      | _ -> mkFix doc Title.Float.Separate.all3 edits |> List.singleton
+
+  /// Removes or adds digit group separators (`_`)
+  let digitGroupFixes doc (lineStr: String) (constant: FloatConstant) =
+    match DigitGroup.removeFix doc lineStr constant.Range constant.ValueRange with
+    | [] -> separateDigitGroupsFix doc lineStr constant
+    | fix -> fix
+
+  let all
+    doc
+    (pos: FcsPos)
+    (lineStr: String)
+    (parseAndCheck: ParseAndCheckResults)
+    (error: bool)
+    (constant: FloatConstant)
+    =
+    [ if not error then
+        // Note: `infinity` & co don't get parsed as `SynConst`, but instead as `Ident`
+        //       -> `constant` is always actual float value, not named
+        yield!
+          CommonFixes.replaceFloatWithNameFix
+            doc
+            pos
+            lineStr
+            parseAndCheck
+            constant.Constant
+            constant.Range
+            constant.Value
+
+      yield! digitGroupFixes doc lineStr constant
+
+      if DEBUG then
+        debugFix doc lineStr constant ]
+
+
+/// CodeFixes for number-based Constant to:
+/// * Convert between bases & forms
+/// * Add digit group separators
+/// * Replace with name (like `infinity` or `TYPE.MinValue`)
+/// * Integrate/Extract Minus (Hex/Oct/Bin -> sign bit vs. explicit `-` sign)
+let fix (getParseResultsForFile: GetParseResultsForFile) : CodeFix =
+  fun (codeActionParams) ->
+    asyncResult {
+      let filePath = codeActionParams.TextDocument.GetFilePath() |> Utils.normalizePath
+      let fcsPos = protocolPosToPos codeActionParams.Range.Start
+      let! (parseAndCheck, lineStr, sourceText) = getParseResultsForFile filePath fcsPos
```

Comment:
One thing I'm thinking about here - could this codefix be authored without the `ParseAndCheckResults`, instead relying only on the `ParseResults`? Because that would be a lot less work to compute.

Reply:
I looked a bit ahead and it's not _strictly_ possible, since the actual fix work requires data on the CheckResults. It may be useful for us to think about this more - ideally we'd be able to very quickly check if certain codefixes are even viable based on parse results as much as possible, then only if they might be viable start using CheckResults.

No changes required for this PR, just something that is on my mind as I review it.

Reply:
We basically had this same idea recently in F# analyzers, could be solved in similar way.

Reply:
> One thing I'm thinking about here - could this codefix be authored without the ParseAndCheckResults, instead relying only on the ParseResults? Because that would be a lot less work to compute.

Aren't `ParseResults` and `CheckResults` calculated and then cached at the same time? So as long as we collect both once we already have them "for free" everywhere else? And because we want diagnostics we already always need both. So `ParseAndCheckResults` is already collected before each CodeFix gets called.

Reply:
Currently yes, I believe you're right - I'm thinking more future-facing. I think we will eventually need some more granular data-dependency-based system to make sure we're not doing work unnecessarily.